### PR TITLE
Bump Eclipse to e4.30 and Tycho to 4.0.4

### PR DESCRIPTION
**Pull request by cmark**
This PR bumps the target platform to Eclipse Platform 4.30 and Tycho to 4.0.4.

The following dependency changes have been applied:
* Upgrade Elasticsearch to 8.12.1
* Upgrade Xtext to 2.33.0
* Upgrade EMF to 2.35.0
* Upgrade MWE2 to 2.16.0
* Upgrade ECL to 2.1.5
* Upgrade SLF4J to 2.0.11
* Upgrade Logback to 1.4.14
* Upgrade Jackson to 2.16.1
* Upgrade Jetty to 12.0.3
* Upgrade to Jakarta EE10
* Upgrade Spring to 6.1.3
* Upgrade springdoc to 2.3.0
* Upgrade swagger to 2.2.20
* Upgrade Guava to 32.1.3-jre
* Upgrade OWLAPI to 4.5.26
* Upgrade Protege to custom 5.0.7
* Upgrade snomed-owl-toolkit to a custom 5.1.1
* Upgrade Fastutil to 8.5.12
* Upgrade RestAssured to 4.5.1
* Upgrade Mockito to 5.9.0
* Upgrade ByteBuddy to 1.14.11
* Upgrade uuid-creator to 5.3.7
* Upgrade failsafe to 3.3.2
* Upgrade rxjava to 2.2.21
* Upgrade netty to 4.1.106.Final
* Upgrade kotlin-osgi-bundle to 1.9.22
* Upgrade mapdb to 3.1.0
* Upgrade failsafe to 3.2.2
* Upgrade micrometer to 1.12.2
* Upgrade zjsonpatch to 0.4.16
* Upgrade bucket4j to 8.7.0
* Upgrade hibernate validator to 8.0.1.Final
* Upgrade java-jwt to 4.4.0
* Upgrade jwks-rsa to 0.22.1
* Upgrade testcontainers to 1.19.4
* Add required Apache Commons libraries.

Build changes:
* Add mvn wrapper
* Tycho is now 4.0.4
* jacoco is now 0.8.11
* Remove unused jars from the produced packages

Jira: https://snowowl.atlassian.net/browse/SO-5973

### CODE REVIEW

File: cis/com.b2international.snowowl.snomed.cis.product/assembly/common/bin/cis.sh
Comment (apeteri) [inline cis/com.b2international.snowowl.snomed.cis.product/assembly/common/bin/cis.sh]: If `-Dosgi.console` works, it can be removed from this line (`cis.bat` is already updated this way):

```suggestion
  -jar plugins/org.eclipse.equinox.launcher_1.6.600.v20231106-1826.jar
```

File: commons/com.b2international.collections.jackson/src/com/b2international/collections/deser/CollectionDeserializers.java
Comment (apeteri) [inline commons/com.b2international.collections.jackson/src/com/b2international/collections/deser/CollectionDeserializers.java]: I guess we can't exactly expect two things in this exception, but IMO `JsonToken.VALUE_NUMBER_FLOAT` would be a tiny bit more appropriate here as the main use case is deserialization of arrays containing floats.

File: commons/com.b2international.index.test.tools/elasticsearch-default-memory-vm.options
Comment (apeteri) [inline commons/com.b2international.index.test.tools/elasticsearch-default-memory-vm.options:2]: Impressive numbers! 😄 Is there a special meaning attached to `Integer.MAX_VALUE`?

File: core/com.b2international.snowowl.core.rest.tests/pom.xml
Comment (apeteri) [inline core/com.b2international.snowowl.core.rest.tests/pom.xml]: Is this test going to remain skipped?

File: core/com.b2international.snowowl.core.rest/.classpath
Comment (apeteri) [inline core/com.b2international.snowowl.core.rest/.classpath]: ⚠️ Source attachment on classpath entry points to local path

File: core/com.b2international.snowowl.core.rest/pom.xml
Comment (apeteri) [inline core/com.b2international.snowowl.core.rest/pom.xml]: How about this cleaning step? Should it remain skipped?

File: core/com.b2international.snowowl.core.rest/src/com/b2international/snowowl/core/rest/SnowOwlApiConfig.java
Comment (apeteri) [inline core/com.b2international.snowowl.core.rest/src/com/b2international/snowowl/core/rest/SnowOwlApiConfig.java:228]: Does the introduction of this `static` modifier have any significance? Are you calling it outside Spring's DI?

File: core/com.b2international.snowowl.core.rest/src/com/b2international/snowowl/core/rest/SnowOwlApiConfig.java
Comment (apeteri) [inline core/com.b2international.snowowl.core.rest/src/com/b2international/snowowl/core/rest/SnowOwlApiConfig.java]: Looks like this is a Servlet-level configuration parameter now:

> Note: In order to use Servlet container based multipart parsing, you need to mark the affected servlet with a "multipart-config" section in web.xml, or with a `MultipartConfigElement` in programmatic servlet registration, or (in case of a custom servlet class) possibly with a `MultipartConfig` annotation on your servlet class. Configuration settings such as maximum sizes or storage locations need to be applied at that servlet registration level; a Servlet container does not allow for them to be set at the `MultipartResolver` level. 

* [Upgrading to Spring Framework 6.x](https://github.com/spring-projects/spring-framework/wiki/Upgrading-to-Spring-Framework-6.x#web-applications-1)
* Javadoc of [StandardServletMultipartResolver](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/web/multipart/support/StandardServletMultipartResolver.html) (quote source)

Comment (apeteri) [reply core/com.b2international.snowowl.core.rest/src/com/b2international/snowowl/core/rest/SnowOwlApiConfig.java]: I think this also means that the method can be removed entirely.

File: snomed/com.b2international.snowowl.snomed.core.rest.tests/pom.xml
Comment (apeteri) [inline snomed/com.b2international.snowowl.snomed.core.rest.tests/pom.xml]: Also this one

File: releng/com.b2international.snowowl.server.update/snow-owl-oss.product
Comment (apeteri) [inline releng/com.b2international.snowowl.server.update/snow-owl-oss.product:4]: Tycho does not seem to support `autoIncludeRequirements` yet (at least the issue is still open): https://github.com/eclipse-tycho/tycho/issues/1056

Do we need this option, is it helping with something within the IDE?

File: fhir/com.b2international.snowowl.fhir.rest.tests/pom.xml
Comment (apeteri) [inline fhir/com.b2international.snowowl.fhir.rest.tests/pom.xml]: And this one. 😸

**Review by apeteri — CHANGES_REQUESTED** [review]
Looks great at first glance! See feedback below.

File: commons/com.b2international.index.test.tools/elasticsearch-default-memory-vm.options
Comment (cmark) [reply commons/com.b2international.index.test.tools/elasticsearch-default-memory-vm.options:2]: I copied this from the `testcontainers:elasticsearch` module because it was unable to load this file from its jar during runtime in OSGi context. See the original file [here](https://github.com/testcontainers/testcontainers-java/blob/main/modules/elasticsearch/src/main/resources/elasticsearch-default-memory-vm.options)

File: core/com.b2international.snowowl.core.rest.tests/pom.xml
Comment (cmark) [reply core/com.b2international.snowowl.core.rest.tests/pom.xml]: I think tests also fail due to the same Jetty 12 issue, but need to verify this on Linux machines as well. I can enable them temporarily and see if they pass in the GitHub Action env.

File: core/com.b2international.snowowl.core.rest/.classpath
Comment (cmark) [reply core/com.b2international.snowowl.core.rest/.classpath]: 😭

File: commons/com.b2international.index.test.tools/elasticsearch-default-memory-vm.options
Comment (apeteri) [reply commons/com.b2international.index.test.tools/elasticsearch-default-memory-vm.options:2]: It's in bytes, didn't notice this at first glance. So 2 GB.

File: core/com.b2international.snowowl.core.rest/pom.xml
Comment (cmark) [reply core/com.b2international.snowowl.core.rest/pom.xml]: No, this was only to prevent failures in local builds running inside Eclipse. Thanks!

File: core/com.b2international.snowowl.core.rest/src/com/b2international/snowowl/core/rest/SnowOwlApiConfig.java
Comment (cmark) [reply core/com.b2international.snowowl.core.rest/src/com/b2international/snowowl/core/rest/SnowOwlApiConfig.java:228]: No, we do not call this outside Spring's DI. Spring while boots prints this warning message to the console if this method is not static.
```
[2024-02-09T13:20:14.299] WARN  main  s.c.s.PostProcessorRegistrationDelegate$BeanPostProcessorChecker Bean 'snowOwlApiConfig' of type [com.b2international.snowowl.core.rest.SnowOwlApiConfig$$SpringCGLIB$$0] is not eligible for getting processed by all BeanPostProcessors (for example: not eligible for auto-proxying). The currently created BeanPostProcessor [methodValidationPostProcessor] is declared through a non-static factory method on that class; consider declaring it as static instead. 
```

File: core/com.b2international.snowowl.core.rest/src/com/b2international/snowowl/core/rest/SnowOwlApiConfig.java
Comment (cmark) [reply core/com.b2international.snowowl.core.rest/src/com/b2international/snowowl/core/rest/SnowOwlApiConfig.java]: Yes, thanks. I haven't got enough time to dig into this yet, but will do so now before merging the PR.

File: releng/com.b2international.snowowl.server.update/snow-owl-oss.product
Comment (cmark) [reply releng/com.b2international.snowowl.server.update/snow-owl-oss.product:4]: I did not explicitly set this, probably it was the Eclipse editor updating the stuff behind the scenes. Not sure if we need it or not. The produced product looks good actually.

File: fhir/com.b2international.snowowl.fhir.rest.tests/pom.xml
Comment (cmark) [reply fhir/com.b2international.snowowl.fhir.rest.tests/pom.xml]: Yes, the three rest api tests failed on my computer during the maven build. I'll enable them now to see if they pass on our Unix build platform.

File: commons/com.b2international.index.test.tools/elasticsearch-default-memory-vm.options
Comment (cmark) [reply commons/com.b2international.index.test.tools/elasticsearch-default-memory-vm.options:2]: Yes :)

File: README.md
Comment (nagyo) [inline README.md]: This could already be 2023-12

File: README.md
Comment (nagyo) [inline README.md]: With the eclipse version upgrade, you can use the e4.30 update site

File: README.md
Comment (nagyo) [inline README.md]: Also, this one is 2.33.0, isn't it?

File: cis/com.b2international.snowowl.snomed.cis.product/assembly/common/bin/cis.sh
Comment (nagyo) [inline cis/com.b2international.snowowl.snomed.cis.product/assembly/common/bin/cis.sh:76]: Does the removal of the -console parameter come from the fact that the default is the same? We still need the osgi console capability in both the CIS and the OSS product.

File: core/com.b2international.snowowl.core.rest.tests/src/configuration/jetty-deploy.xml
Comment (nagyo) [inline core/com.b2international.snowowl.core.rest.tests/src/configuration/jetty-deploy.xml:1]: I thought we wouldn't need to copy these files, but I might be outdated with the jetty and osgi boot issue?

File: cis/com.b2international.snowowl.snomed.cis.product/assembly/common/bin/cis.sh
Comment (nagyo) [reply cis/com.b2international.snowowl.snomed.cis.product/assembly/common/bin/cis.sh:76]: I see that we are not using that anymore in the OSS product, I guess it is a historical residue 👍🏻

**Review by nagyo — APPROVED** [review]
LGTM! (as much as I was able to go through all the files 😅)

File: README.md
Comment (cmark) [reply README.md]: Right, this should be 2.33.0.

File: cis/com.b2international.snowowl.snomed.cis.product/assembly/common/bin/cis.sh
Comment (cmark) [reply cis/com.b2international.snowowl.snomed.cis.product/assembly/common/bin/cis.sh:76]: Yes

File: core/com.b2international.snowowl.core.rest.tests/src/configuration/jetty-deploy.xml
Comment (cmark) [reply core/com.b2international.snowowl.core.rest.tests/src/configuration/jetty-deploy.xml:1]: There is another issue with osgi.boot on Jetty 12. The only workaround now is to use explicit jetty files in all our e2e tests.

File: README.md
Comment (apeteri) [inline README.md]: Now it's 2023-09 and 2023-12 at the same time.

**Review by apeteri — APPROVED** [review]
OK to merge once the README is adjusted and (possibly) the multipart upload configuration is added back to its new place.

File: README.md
Comment (cmark) [reply README.md]: Fixed

File: core/com.b2international.snowowl.core.rest/src/com/b2international/snowowl/core/rest/SnowOwlApiConfig.java
Comment (cmark) [reply core/com.b2international.snowowl.core.rest/src/com/b2international/snowowl/core/rest/SnowOwlApiConfig.java]: Done

**Pull request merged**
